hypothesis/h
# Add eLife annotation timestamps data migration

We're going to run an API script to import 1000+ Disqus comments from eLife into page note annotations using the standard annotation create API. After being created in this way the annotations will have the wrong timestamps - they'll be the times at which the annotations were created in h, not the times of the comments exported from Disqus. This migration sets the timestamps of these imported annotations to their correct Disqus-exported timestamps.

To test, checkout this branch and then:

1. Create a user account in your h dev instance, and get a developer API token for this account

2. Create a group in your h dev instance

3. Set the `IMPORT_GROUP_ID` env var to the `groupid` of the group you just created (the `import.py` script below uses this): `export IMPORT_GROUP_ID=<YOUR_GROUP_ID>`

4. You need to have annotations in your DB with `imported_id` extra fields matching those from eLife (in the `TIMESTAMPS` list in the migration script). Here's a [script that will add test annotations with these IDs to your dev instance of h](https://gist.github.com/seanh/09112f2e39a3381fd6a156750e91c789). To run it: `./import.py <YOUR_H_DEVELOPER_API_TOKEN>`.

   At this point the `created` and `updated` times of the imported annotations in your DB will be recent. You can inspect them with: `psql postgresql://postgres@localhost/postgres -c 'SELECT created, updated FROM annotation;'`.

5. Edit the migration script, changing the `GROUPID` constant to be the `groupid` of the group you just created

6. Stamp your db: `./bin/hypothesis migrate stamp 9bcc39244e82`

7. Run the migration script: `./bin/hypothesis migrate upgrade +1`

You should see:

```
[INFO] Running upgrade 9bcc39244e82 -> b32d09564cf5, Correct the timestamps of imported eLife annotations.
[INFO] Updated the timestamps of 1258 annotations
```

And if you inspect your database you should see that the annotation's timestamps have been changed: `psql postgresql://postgres@localhost/postgres -c 'SELECT created, updated FROM annotation;'`.

You can also run the migration script _without_ having imported the annotations into your DB. In that case it should not crash and you should see `Updated the timestamps of 0 annotations`.

If you have a production database dump loaded than you can use the developer API token of the actual disqus import user, and the groupid of the actual eLife group, and you don't need to edit the migration script to test it.
